tests/test_allfeatures.py

```diff
@@ -41,123 +42,82 @@ def get_allfeature_values():
     """Get back all the feature names and value"""
 
     import efel
-    efel.reset()
     import numpy
+    efel.reset()
 
     all_featurenames = efel.getFeatureNames()
```

Comment:
@AurelienJaquier had a nice suggestion for testing all features: Test all features for a trace with no spike and see if they generate the desired results.

Reply:
Oh good idea, we can give a constant voltage that is always -80.0 to see how each feature will handle it.

Reply:
This is revealing issues already.

```
    "ISI_CV": None,
    "ISI_log_slope": None,
    "ISI_log_slope_skip": None,
    "ISI_semilog_slope": None,
    "ISI_values": None,
    "Spikecount": array([0]),
    "Spikecount_stimint": None,
```

Reply:
In this case Spikecount is correct to be 0. Spikecount_stimint should also be 0.

Reply:
Making Spikecount 0 instead of None is really hacky since the dependency resolver will automatically fail Spikecount when "peak_indices" fail. I am now more inclined to think towards producing a None like in the other features.


```
    "APlast_amp": None,
    "APlast_width": None,
    "ISI_CV": None,
    "ISI_log_slope": None,
    "ISI_log_slope_skip": None,
    "ISI_semilog_slope": None,
    "ISI_values": None,
    "Spikecount": None,
    "Spikecount_stimint": None,
    "adaptation_index": None,
    "adaptation_index2": None,
```

Reply:
I added the test that is calling all of the features. There are no longer [] values. There are [0] values but those are scientifically meaningful e.g. `maximum_voltage_from_voltagebase`.

```
    "maximum_voltage": array([-80.0]),
    "maximum_voltage_from_voltagebase": array([0.0]),
```
 The 0.0 here is not the missing value but the real value of the computation.

Reply:
These are all of the features extracted on a trace with a constant -80.0 voltage. Obviously, there are no spikes. The missing value handling looks correct to me. Could you check also @darshanmandge @AurelienJaquier ?

```
{
    "ADP_peak_amplitude": None,
    "ADP_peak_indices": None,
    "ADP_peak_values": None,
    "AHP1_depth_from_peak": None,
    "AHP2_depth_from_peak": None,
    "AHP_depth": None,
    "AHP_depth_abs": None,
    "AHP_depth_abs_slow": None,
    "AHP_depth_diff": None,
    "AHP_depth_from_peak": None,
    "AHP_depth_slow": None,
    "AHP_slow_time": None,
    "AHP_time_from_peak": None,
    "AP1_amp": None,
    "AP1_begin_voltage": None,
    "AP1_begin_width": None,
    "AP1_peak": None,
    "AP1_width": None,
    "AP2_AP1_begin_width_diff": None,
    "AP2_AP1_diff": None,
    "AP2_AP1_peak_diff": None,
    "AP2_amp": None,
    "AP2_begin_voltage": None,
    "AP2_begin_width": None,
    "AP2_peak": None,
    "AP2_width": None,
    "AP_amplitude": None,
    "AP_amplitude_change": None,
    "AP_amplitude_diff": None,
    "AP_amplitude_from_voltagebase": None,
    "AP_begin_indices": None,
    "AP_begin_time": None,
    "AP_begin_voltage": None,
    "AP_begin_width": None,
    "AP_duration": None,
    "AP_duration_change": None,
    "AP_duration_half_width": None,
    "AP_duration_half_width_change": None,
    "AP_end_indices": None,
    "AP_fall_indices": None,
    "AP_fall_rate": None,
    "AP_fall_rate_change": None,
    "AP_fall_time": None,
    "AP_height": None,
    "AP_peak_downstroke": None,
    "AP_peak_upstroke": None,
    "AP_phaseslope": None,
    "AP_rise_indices": None,
    "AP_rise_rate": None,
    "AP_rise_rate_change": None,
    "AP_rise_time": None,
    "AP_width": None,
    "AP_width_between_threshold": None,
    "APlast_amp": None,
    "APlast_width": None,
    "ISI_CV": None,
    "ISI_log_slope": None,
    "ISI_log_slope_skip": None,
    "ISI_semilog_slope": None,
    "ISI_values": None,
    "Spikecount": None,
    "Spikecount_stimint": None,
    "adaptation_index": None,
    "adaptation_index2": None,
    "all_ISI_values": None,
    "amp_drop_first_last": None,
    "amp_drop_first_second": None,
    "amp_drop_second_last": None,
    "burst_ISI_indices": None,
    "burst_begin_indices": None,
    "burst_end_indices": None,
    "burst_mean_freq": None,
    "burst_number": None,
    "current": None,
    "current_base": None,
    "decay_time_constant_after_stim": None,
    "depolarized_base": None,
    "doublet_ISI": None,
    "fast_AHP": None,
    "fast_AHP_change": None,
    "interburst_min_indices": None,
    "interburst_min_values": None,
    "interburst_voltage": None,
    "inv_fifth_ISI": None,
    "inv_first_ISI": None,
    "inv_fourth_ISI": None,
    "inv_last_ISI": None,
    "inv_second_ISI": None,
    "inv_third_ISI": None,
    "inv_time_to_first_spike": None,
    "irregularity_index": None,
    "is_not_stuck": None,
    "max_amp_difference": None,
    "maximum_voltage": array([-80.0]),
    "maximum_voltage_from_voltagebase": array([0.0]),
    "mean_AP_amplitude": None,
    "mean_frequency": None,
    "min_AHP_indices": None,
    "min_AHP_values": None,
    "min_between_peaks_indices": None,
    "min_between_peaks_values": None,
    "min_voltage_between_spikes": None,
    "minimum_voltage": array([-80.0]),
    "multiple_decay_time_constant_after_stim": None,
    "number_initial_spikes": None,
    "ohmic_input_resistance": None,
    "ohmic_input_resistance_vb_ssse": None,
    "peak_indices": None,
    "peak_time": None,
    "peak_voltage": None,
    "postburst_min_indices": None,
    "postburst_min_values": None,
    "sag_amplitude": array([0.0]),
    "sag_ratio1": None,
    "sag_ratio2": None,
    "sag_time_constant": None,
    "single_burst_ratio": None,
    "spike_half_width": None,
    "spike_width2": None,
    "steady_state_hyper": array([-80.0]),
    "steady_state_voltage": array([-80.0]),
    "steady_state_voltage_stimend": array([-80.0]),
    "strict_burst_mean_freq": None,
    "strict_burst_number": None,
    "strict_interburst_voltage": None,
    "time": array([0.000e00, 1.000e-01, 2.000e-01, ..., 9.988e02, 9.989e02, 9.990e02]),
    "time_constant": None,
    "time_to_first_spike": None,
    "time_to_interburst_min": None,
    "time_to_last_spike": None,
    "time_to_second_spike": None,
    "trace_check": None,
    "voltage": array([-80.0, -80.0, -80.0, ..., -80.0, -80.0, -80.0]),
    "voltage_after_stim": array([-80.0]),
    "voltage_base": array([-80.0]),
    "voltage_deflection": array([0.0]),
    "voltage_deflection_begin": array([0.0]),
    "voltage_deflection_vb_ssse": array([0.0]),
    "ISIs": None,
    "initburst_sahp": None,
    "initburst_sahp_vb": None,
    "initburst_sahp_ssse": None,
    "depol_block": array([1]),
    "depol_block_bool": array([0]),
    "spikes_per_burst": None,
    "spikes_per_burst_diff": None,
    "spikes_in_burst1_burst2_diff": None,
    "spikes_in_burst1_burstlast_diff": None,
    "impedance": None,
}
```

Reply:
Update: `Spikecount` and `Spikecount_stimint` are moved to Python. We can set them to be 0.0 without forcing `peak_indices` to be an empty vector.

Reply:
There are also these features that should be `0`, like `Spikecount`: `burst_number`, `strict_burst_number`, `number_initial_spikes`.
Also the `trace_check` should be `0` here. It should return `0` if there is no spike outside of stimulus, and `None` otherwise. When there is no spike, obviously there is no spike outside of stimulus.
We could set a `stimulus_current` in the setting so that `ohmic_input_resistance` and `ohmic_input_resistance_vb_ssse` return something.
Also I don't get why `decay_time_constant_after_stim` and `multiple_decay_time_constant_after_stim` are `None`. Do you have a stimulus going all the way up to the last datapoint or something?

Reply:
`burst_number` and `strict_burst_number` are now returning 0 in case there is no burst - similar to Spikecount.

The `number_initial_spikes` however is trickier since it fails when the following condition is not met: `"initial_perc should lie between [0 1)."` in that case it returns a `None`. When it does not fail and yet cannot find any spikes then it should return a 0 meaning that there are 0 spikes. I think there are other feature implementations that are in a similar situation. I created this issue https://github.com/BlueBrain/eFEL/issues/341 to discuss it and address it in more detail.

Reply:
Let me update `trace_check` as well. It is currently failing due to a failing dependency: `peak_time`.
It will be a clearer implementation to move `trace_check` from LibV1 to Python. Otherwise it adds complications as it requires bypassing the dependency resolver in C++.

Reply:
Oh, `stimulus_current` is `0` here, right? No wonder `ohmic_input_resistance` is `None`.

Reply:
Yes the stimulus_current is not set at the moment.

Reply:
trace_check now behaves as expected.

Reply:
Updated results upon constant voltage + stimulus current parameter:

```python

feature_values
{
    "ADP_peak_amplitude": None,
    "ADP_peak_indices": None,
    "ADP_peak_values": None,
    "AHP1_depth_from_peak": None,
    "AHP2_depth_from_peak": None,
    "AHP_depth": None,
    "AHP_depth_abs": None,
    "AHP_depth_abs_slow": None,
    "AHP_depth_diff": None,
    "AHP_depth_from_peak": None,
    "AHP_depth_slow": None,
    "AHP_slow_time": None,
    "AHP_time_from_peak": None,
    "AP1_amp": None,
    "AP1_begin_voltage": None,
    "AP1_begin_width": None,
    "AP1_peak": None,
    "AP1_width": None,
    "AP2_AP1_begin_width_diff": None,
    "AP2_AP1_diff": None,
    "AP2_AP1_peak_diff": None,
    "AP2_amp": None,
    "AP2_begin_voltage": None,
    "AP2_begin_width": None,
    "AP2_peak": None,
    "AP2_width": None,
    "AP_amplitude": None,
    "AP_amplitude_change": None,
    "AP_amplitude_diff": None,
    "AP_amplitude_from_voltagebase": None,
    "AP_begin_indices": None,
    "AP_begin_time": None,
    "AP_begin_voltage": None,
    "AP_begin_width": None,
    "AP_duration": None,
    "AP_duration_change": None,
    "AP_duration_half_width": None,
    "AP_duration_half_width_change": None,
    "AP_end_indices": None,
    "AP_fall_indices": None,
    "AP_fall_rate": None,
    "AP_fall_rate_change": None,
    "AP_fall_time": None,
    "AP_height": None,
    "AP_peak_downstroke": None,
    "AP_peak_upstroke": None,
    "AP_phaseslope": None,
    "AP_rise_indices": None,
    "AP_rise_rate": None,
    "AP_rise_rate_change": None,
    "AP_rise_time": None,
    "AP_width": None,
    "AP_width_between_threshold": None,
    "APlast_amp": None,
    "APlast_width": None,
    "ISI_CV": None,
    "ISI_log_slope": None,
    "ISI_log_slope_skip": None,
    "ISI_semilog_slope": None,
    "ISI_values": None,
    "adaptation_index": None,
    "adaptation_index2": None,
    "all_ISI_values": None,
    "amp_drop_first_last": None,
    "amp_drop_first_second": None,
    "amp_drop_second_last": None,
    "burst_ISI_indices": None,
    "burst_begin_indices": None,
    "burst_end_indices": None,
    "burst_mean_freq": None,
    "current": None,
    "current_base": None,
    "decay_time_constant_after_stim": None,
    "depolarized_base": None,
    "doublet_ISI": None,
    "fast_AHP": None,
    "fast_AHP_change": None,
    "interburst_min_indices": None,
    "interburst_min_values": None,
    "interburst_voltage": None,
    "inv_fifth_ISI": None,
    "inv_first_ISI": None,
    "inv_fourth_ISI": None,
    "inv_last_ISI": None,
    "inv_second_ISI": None,
    "inv_third_ISI": None,
    "inv_time_to_first_spike": None,
    "irregularity_index": None,
    "is_not_stuck": None,
    "max_amp_difference": None,
    "maximum_voltage": array([-80.0]),
    "maximum_voltage_from_voltagebase": array([0.0]),
    "mean_AP_amplitude": None,
    "mean_frequency": None,
    "min_AHP_indices": None,
    "min_AHP_values": None,
    "min_between_peaks_indices": None,
    "min_between_peaks_values": None,
    "min_voltage_between_spikes": None,
    "minimum_voltage": array([-80.0]),
    "multiple_decay_time_constant_after_stim": None,
    "number_initial_spikes": None,
    "ohmic_input_resistance": array([0.0]),
    "ohmic_input_resistance_vb_ssse": array([0.0]),
    "peak_indices": None,
    "peak_time": None,
    "peak_voltage": None,
    "postburst_min_indices": None,
    "postburst_min_values": None,
    "sag_amplitude": array([0.0]),
    "sag_ratio1": None,
    "sag_ratio2": None,
    "sag_time_constant": None,
    "single_burst_ratio": None,
    "spike_half_width": None,
    "spike_width2": None,
    "steady_state_hyper": array([-80.0]),
    "steady_state_voltage": array([-80.0]),
    "steady_state_voltage_stimend": array([-80.0]),
    "strict_burst_mean_freq": None,
    "strict_interburst_voltage": None,
    "time": array([0.000e00, 1.000e-01, 2.000e-01, ..., 9.988e02, 9.989e02, 9.990e02]),
    "time_constant": None,
    "time_to_first_spike": None,
    "time_to_interburst_min": None,
    "time_to_last_spike": None,
    "time_to_second_spike": None,
    "voltage": array([-80.0, -80.0, -80.0, ..., -80.0, -80.0, -80.0]),
    "voltage_after_stim": array([-80.0]),
    "voltage_base": array([-80.0]),
    "voltage_deflection": array([0.0]),
    "voltage_deflection_begin": array([0.0]),
    "voltage_deflection_vb_ssse": array([0.0]),
    "ISIs": None,
    "initburst_sahp": None,
    "initburst_sahp_vb": None,
    "initburst_sahp_ssse": None,
    "depol_block": array([1]),
    "depol_block_bool": array([0]),
    "Spikecount": array([0]),
    "Spikecount_stimint": array([0]),
    "spikes_per_burst": None,
    "spikes_per_burst_diff": None,
    "spikes_in_burst1_burst2_diff": None,
    "spikes_in_burst1_burstlast_diff": None,
    "impedance": None,
    "burst_number": array([0]),
    "strict_burst_number": array([0]),
    "trace_check": array([0]),
}

```

Reply:
@AurelienJaquier decay_time_constant fails because of the following:
```C++
if (decayTimes.size() < 1 || decayValues.size() < 1)
    throw FeatureComputationError("No data points to calculate decay_time_constant_after_stim");
```
Decay times and decay values are both empty for that trace.

Reply:
about stimulus current, shouldn't it be 0 for this trace? You don't inject any current, right? If so, maybe reverting it to 0 would be better, even if it means having ohmic_resistance being None as a result

Reply:
If I give 0, the ohmic_resistance will be None to prevent division by 0. Ok sure I can revert